src/mmds/src/token.rs

```diff
@@ -0,0 +1,562 @@
+    /// cannot be decrypted. If decryption succeeds, returns true if token has not expired
+    /// (i.e. current time is greater than expiry) and false otherwise.
+    pub fn is_valid(&self, encoded_token: &str) -> bool {
+        // Check size of encoded token struct.
+        if encoded_token.len() > TOKEN_LENGTH_LIMIT {
+            return false;
+        }
+
+        // Decode token struct from base64.
+        let mut token = match Token::base64_decode_token(encoded_token) {
+            Ok(token) => token,
+            Err(_) => return false,
+        };
+
+        // Validate constant fields inside decoded token structure.
+        if !token.sanity_check() {
+            return false;
+        }
+
+        // Decrypt ttl using AES-GCM block cipher.
+        let expiry = match self.decrypt_expiry(&mut token.payload, &token.tag, &token.iv) {
+            Ok(expiry) => expiry,
+            Err(_) => return false,
+        };
+
+        // Get current time.
+        let time_now = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("System time is before UNIX_EPOCH.");
+        let time_now_as_seconds = time_now.as_secs();
+
+        expiry > time_now_as_seconds
+    }
+
+    /// Decrypt ciphertext composed of payload and tag to obtain the expiry value.
+    fn decrypt_expiry(
+        &self,
+        payload: &mut [u8; PAYLOAD_LEN],
+        tag: &[u8],
+        iv: &[u8],
+    ) -> Result<u64, Error> {
+        // Create Nonce object from initialization vector.
+        let nonce = Nonce::from_slice(iv);
+        // Decrypt expiry as vector of bytes from ciphertext.
+        self.cipher
+            .decrypt_in_place_detached(nonce, b"", payload, aes_gcm::Tag::from_slice(tag))
+            .map_err(|_| Error::ExpiryExtraction)?;
+        let expiry_as_bytes = payload[..]
+            .try_into()
+            .map_err(|_| Error::ExpiryExtraction)?;
+
+        // Return expiry value in seconds.
+        Ok(u64::from_le_bytes(expiry_as_bytes))
+    }
+
+    /// Create a new AES-GCM cipher entity.
+    fn create_cipher() -> Result<Aes256Gcm, Error> {
+        // Randomly generate a 256-bit key to be used for encryption/decryption purposes.
+        let mut key = [0u8; KEY_LEN];
+        read_into_buf_from_entropy_pool(&mut key).map_err(Error::EntropyPool)?;
+
+        // Create cipher entity to handle encryption/decryption.
+        Ok(Aes256Gcm::new(Key::from_slice(&key)))
+    }
+
+    /// Make sure to reinitialize the cipher under a new key before reaching
+    /// a count of 2^32 encrypted tokens under the same cipher entity.
+    fn check_encryption_count(&mut self) -> Result<(), Error> {
+        // Make sure no more than 2^32 - 1 tokens are encrypted under
+        // the same encryption key.
+        if self.num_encrypted_tokens == u32::MAX {
+            // Reinitialize the cipher entity under a new key.
+            self.cipher = TokenAuthority::create_cipher()?;
```

Comment:
I would specify somewhere the fact that by design we don't retain the old cipher because this case shouldn't happen and if it happens we expect the customer code to implement a retry when the token is expired.

Reply:
Done.